js/command.ts

```diff
@@ -0,0 +1,193 @@
+// Copyright 2018 the Deno authors. All rights reserved. MIT license.
+import * as dispatch from "./dispatch";
+import { DenoError, ErrorKind } from "./errors";
+import * as flatbuffers from "./flatbuffers";
+import * as msg from "gen/msg_generated";
+import { assert, isObject, unreachable } from "./util";
+
+export interface CommandOptions {
+  argv?: string[];
+  dir?: string | null;
+  throw?: boolean;
+}
+
+export interface Command extends CommandOptions {
+  argv: [string, ...string[]];
+}
+
+interface ExitCode {
+  success: boolean;
+  code: number;
+  signal: null;
+}
+
+interface ExitSignal {
+  success: false;
+  code: null;
+  signal: number; // TODO: Make this a string, e.g. 'SIGTERM'.
+}
+
+export type ExitStatus = ExitCode | ExitSignal;
+
+// TODO: Move some of these small helper types to types.ts. Typescript won't
+// allow it for some reason: "Import or export declaration in an ambient module
+// declaration cannot reference module through relative module name."
+
+// Exclude from T those types that are assignable to U.
+export type Exclude<T, U> = T extends U ? never : T;
+// An array or arraylike type with at least one element.
+export type NonEmptyList<T> = [T, ...T[]];
+// Remove the 'optional' modifier from all properties of T.
+export type Required<T> = { [P in keyof T]-?: T[P] };
+
+// Make a tuple type longer by prepending an element type.
+// Unfortunately `type Prepend<S, T extends Array> = [S, ...T[]]` doesn't work.
+// prettier-ignore
+type Prepend<S, T extends Array<unknown>> =
+  ((first: S, ...rest: T) => void) extends ((...all: infer R) => void)
+    ? R
+    : never;
+
+// This type helps us define the call signature for `run()`, which accepts a
+// variable number of arguments followed by an optional options object.
```

Comment:
If that doesn't work out then we can fall back on the signature of the implementation which is `run(...args: Array<string | options>)`

Reply:
> This feels superfluous to me... It's all to save two characters?

It was driven by the motivation to get as convenient as python more or better.
In node you do this: `execFile('git', ["commit", "src/*.ts", "-m", '"works!'"], { cwd: 'src/deno', env: { ...env, GIT_AUTHOR_NAME: 'me' }}})` 

I want fewer brackets and braces.

Reply:
While it former works from a type perspective, it is a bit incoherent from an intellisense perspective.  I would argue both are ergonomic.  Would it make sense to just have two overloads, either the "straight forward" or the options bag:

```ts
run(...args: string[]): Promise<ExitStatus>;
run(options: Opts): Promise<ExitStatus>;
```

Where `args` is required... so:

```ts
run("python", "-c", "print 'hello'");
// or
run({
  args: ["python", "-c", "print 'hello'"],
  dir: "/"
});
```

Reply:
> Would it make sense to just have two overloads, either the "straight forward" or the options bag:

I may do that indeed, although I think that I might as well just revert to doing just the options bag -- setting some options is in my experience pretty much always required.

Btw - I intended `run` to be a convenience API. The low level one will probably only take an options bag. But I had to can that because the low level api will just start a command and return a process object -- not run it to completion. But setting up the associated resource on the rust side was more difficult than I expected, so I punt on that for now.

Other ideas:

* We could do `run(...argv_and_options: Array<string | Options>)` and merge the options objects if there are multiple.
  A nice property may be that if you want to run a number of commands you can create a template `options` object and easily apply it to all commands. Like
    ```
  let defaults = { 
    cwd: '/my/src/deno',
    env: { PYTHONPATH: '/my/weird/python' }
  }
  run("python", "tools/build.py", defaults)
  run("git", "commit", "-m", "identity theft", defaults, { env: { GIT_AUTHOR_NAME: "Ryan Dahl" }});
  ```
  
  or even:
  
  ```
  let git = {
    cwd: '/my/src/deno',
    argv: ['git'],
  }
  run(git, "log");
  run(git, "commit", "-m", "bla");
  ```
  
  If the user puts the options object in the middle like `run("python", "-c", { cwd: "/there" }, "hello")` , then, well that's weird but why would we care?

* Another would be to use a builder which is common in strongly typed languages. The api then becomes something like `command("python", "-c", "hello").cwd("/there").run()`. I feel it's a bit too high level maybe, and also pretty verbose for casual use. And although it would work now, when the time comes to add exotic options like `setgid` and `run_in_a_new_window_on_a_hidden_desktop` and whatnot we'd have to add all these methods.

Reply:
@piscisaureus Regarding setgid, it is done thru [Cmd.SysProcAttr](https://golang.org/pkg/os/exec/#Cmd) 

Regarding the builder, isn't exposing the struct simpler? I find Rust's builder APIs rather annoying. I think we could just have some public members in the `Cmd` class (or interface).